### ISSUE TITLE
Eta reduction suggestions may not be valid due to simplified subsumption.

### ISSUE THREAD

**Issue by galenhuntington** [ndmitchell/hlint#1107]
The simplified subsumption proposal (https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0287-simplify-subsumption.rst) has been implemented in GHC 9.0.  Consider this program:

```haskell
{-# LANGUAGE RankNTypes #-}

g :: forall a. () -> a
g x = undefined

f :: () -> forall a. a
f x = g x
```

`hlint` will recommend eta-reducing the latter definition to `f = g`.  But while this is fine under GHC 8.10, it fails under GHC 9:

```
    • Couldn't match type ‘a0’ with ‘forall a. a’
      Expected: () -> forall a. a
        Actual: () -> a0
      Cannot instantiate unification variable ‘a0’
      with a type involving polytypes: forall a. a
    • In the expression: g
      In an equation for ‘f’: f = g
```

I ran into this problem in https://github.com/ekmett/profunctors/pull/91.  (Yes, I've been PRing GHC 9 updates all day.)  See https://travis-ci.org/github/ekmett/profunctors/jobs/721218651#L667.

Alas, this seems like it might be pretty tricky to detect.

**Comment by googleson78** [comment]
Really tricky, considering hlint works without type-checking.. :(

**Comment by galenhuntington** [comment]
That does indeed make this look rather hopeless.

**Comment by galenhuntington** [comment]
Also, it would've been a better example to swap those definitions, as the above error can in fact be avoided with `-XImpredicativeTypes` (although that's far from obvious).  With this—

```haskell
f :: () -> forall a. a
f = undefined
g :: forall a. () -> a
g = f
```

—we get this error under GHC 9—

```
    • Couldn't match type ‘a’ with ‘forall a1. a1’
      Expected: () -> a
        Actual: () -> forall a. a
      ‘a’ is a rigid type variable bound by
        the type signature for:
          g :: forall a. () -> a
        at test.hs:6:1-22
    • In the expression: f
      In an equation for ‘g’: g = f
```

—and impredicativity won't save us.

**Comment by ndmitchell** [comment]
HLint has always been a bit fast and lose with eta reduction for RankN types, because as the rules in GHC have shifted many times, they've always required some signatures in some places. My guess is this breakage is pretty rare, and there's no way to detect it without types (which we don't have), so my inclination is to slightly strengthen the warning around RankN types in the README at https://github.com/ndmitchell/hlint#bugs-and-limitations and leave it at that.

**Comment by RyanGlScott** [comment]
> My guess is this breakage is pretty rare

Unfortunately, this has not been my experience. On [`head.hackage`](https://gitlab.haskell.org/ghc/head.hackage/) alone, I've had to eta expand over 40 libraries to accommodate simplified subsumption. I suspect that there are many examples than what are in `head.hackage`.

**Comment by ndmitchell** [comment]
I guess the alternative is to just stop hinting for eta-reduce? Or are there any good intermediate states?

**Comment by RyanGlScott** [comment]
I can't speak for others, but my personal inclination would be to just stop hinting about eta reduction. Aside from the issues documented here and in #514, there are also some corner cases where eta expansion (or lack thereof) can negatively affect GHC's inliner. For these reasons, I usually disable `hlint`'s eta reduction hints wholesale.

**Comment by ndmitchell** [comment]
Making it default-off is certainly plausible. There are a few levels of eta reduction - e.g. `f x = h x` and separately `\x y -> x + y` to `(+)`. Would you eliminate all eta reduction? Or just some forms?

**Comment by RyanGlScott** [comment]
> Making it default-off is certainly plausible.

That would be fine with me. In my mind, eta reducing everything is like GCC's `-ffast-math` flag: it can be useful in certain situations, but dangerous if used recklessly. Eta reduction warnings ought to be something that you make a conscious decision to use.

> Would you eliminate all eta reduction? Or just some forms?

I suppose you could employ some heuristics to guess when GHC would be OK with eta reduction and only warn about those, but that seems like it would be brittle. My personal inclination is to be uniform and either warn about everything or nothing.

**Comment by ndmitchell** [comment]
I think we have two broad categories:

1. Eta reduction for eta reduction sake `f x = g x`.
2. Eta reduction to avoid a lambda, `(\x -> x + 1)`.

I'm inclined to keep the second, but disable the first, since I imagine that's more useful (you are fundamentally simplifying the code) but disable the first.

**Comment by RyanGlScott** [comment]
I'm unclear as to why these should be treated differently. The simplified subsumption issue mentioned in https://github.com/ndmitchell/hlint/issues/1107#issue-686003272 applies regardless of whether you're binding variables with a function clause or a lambda. You could just as well write the program in https://github.com/ndmitchell/hlint/issues/1107#issue-686003272 like this:

```hs
{-# LANGUAGE RankNTypes #-}

g :: forall a. () -> a
g x = undefined

f :: () -> forall a. a
f = (\x -> g x)
```

Eta-reducing `(\x -> g x)` to `g` will still ultimately lead to a type error on GHC 9.0+.

**Comment by ndmitchell** [comment]
There are many corner cases whether HLint can introduce an error, so the aim isn't to eliminate all of them, just make them mostly unlikely, while at the same time enabling as many beneficial suggestions as possible. Therefore, there's no absolute line.

**Comment by RyanGlScott** [comment]
> the aim isn't to eliminate all of them, just make them mostly unlikely

Right. I suppose I'm just failing to see why eta reducing lambdas would be fundamentally more likely to work than eta reducing function clauses. As one data point, in [this `head.hackage` patch](https://gitlab.haskell.org/ghc/head.hackage/-/commit/a89f0189a628e5de9a6e8fe9830e29a3e61e316c) which modified Hackage libraries to work with simplified subsumption, about 16 of the 42 libraries needed to be fixed by eta expanding a lambda expression. So while more libraries had function clauses that broke under simplified subsumption, it's not by a substantial amount.

**Comment by georgefst** [comment]
I came across this while looking to see if the following hints had previously been suggested (which would presumably have similar issues re. subsumption):
- `f x = g (h x)` -> `f = g . h`
- `(\f -> f x)` -> `($ x)`

I think it would be a shame to lose the eta-reduction hints in general - for one thing, they often seem to pop out at the end of some other refactoring. Would it really be that much of an inconvenience to disable them for individual functions, or at the module/package/project level for codebases which make significant use of `RankNTypes` etc.? My reading of the Quick Look paper suggests perhaps not.

> I suppose I'm just failing to see why eta reducing lambdas would be fundamentally more likely to work than eta reducing function clauses.

I don't think he was saying that they are more likely to work. Just that because those hints are more useful, the occasional false positive is more worth it.

**Comment by galenhuntington** [comment]
I am also coming around to the idea of leaving this as is, with perhaps some documentation of this new pitfall.  If code includes a pragma annotation—

```haskell
{-# ANN g "HLint: ignore Eta reduce" #-}  
g :: forall a. () -> a           
g x = f x  
```

—it doubles as documentation for humans, who might also be tempted to eta-reduce.

And (per the previous comment) it can always be turned off more broadly for higher-rank-heavy code.
